# Format of probe file

The probe file is generated by *generate_oligos* program, all the raw oligonucletides or probes will be stored in this file and compacted in bgzip format. The probe file consist of *header* and *body* part. All the lines in the *header* part are inited with '#', all the design-related paramenters will be kept in the *header*, and oligonucletides and related information will be kept in the *body* part.

Here is a demo of probe file.
```
##filetype=probe
##generate_oligos Version=0.02
##project_name=LUNG_COLORECTUM
##max_length=90
##Command=-project LUNG_COLORECTUM -l 0 -r /ifs5/ST_TRANS_CARDIO/PUB/analysis_pipelines/HPC_chip/db/aln_db/hg19/hg19_UCSC.fa.gz -u /hwfssz1/ST_MCHRI/CLINIC/PMO/ProbeDesign_workdir/d
#chrom  start   end     seq_length      sequence        n_block starts  ends    repeat_ratio    GC_content      rank
chr7    1586538 1586628 90      GAGGCCTGGGCAGCCTGGGTCCCTACAGCACTGGCAGCCCAGGCCCCCCTACAGGTCCTCCGAGGGGATCAGCATCCGCTTCTCCAGGCT      1       1586538,        1586628,        0.00    0.69
chr7    1586561 1586651 90      TACAGCACTGGCAGCCCAGGCCCCCCTACAGGTCCTCCGAGGGGATCAGCATCCGCTTCTCCAGGCTCCGGCTCTTCCTGCTCCCGCCGG      1       1586561,        1586651,        0.00    0.69
chr7    1586606 1586696 90      TCAGCATCCGCTTCTCCAGGCTCCGGCTCTTCCTGCTCCCGCCGGAGCCGCCGCTGGGGTGGGTGCCGGGCCTGGGCGCCTCGTGCGTGG      1       1586606,        1586696,        0.00    0.76
chr7    1586651 1586741 90      AGCCGCCGCTGGGGTGGGTGCCGGGCCTGGGCGCCTCGTGCGTGGCCTGCTGCGTGTAGTGCTGGTAGGCGGGGGAGAAGTTGTGGATGG      1       1586651,        1586741,        0.00    0.72
chr7    1586696 1586786 90      CCTGCTGCGTGTAGTGCTGGTAGGCGGGGGAGAAGTTGTGGATGGCGTCCTGCACGATGTCCTGGGGGCTCACTGTCTCCCTGATGCCGC      1       1586696,        1586786,        0.00    0.64
chr7    1586741 1586831 90      CGTCCTGCACGATGTCCTGGGGGCTCACTGTCTCCCTGATGCCGCTGGAGATGCTCTGCATGGGTGCCGGGGGGGCTGGGGGAGGGCAGT      1       1586741,        1586831,        0.00    0.69
chr7    1586775 1586865 90      CCTGATGCCGCTGGAGATGCTCTGCATGGGTGCCGGGGGGGCTGGGGGAGGGCAGTGTATGAGCCCCACCATCCCCCCTGCACTCCAGCT      1       1586775,        1586865,        0.00    0.69
chr7    1586775 1586865 90      CCTGATGCCGCTGGAGATGCTCTGCATGGGTGCCGGGGGGGCTGGGGGAGGGCAGTGTATGAGCCCCACCATCCCCCCTGCACTCCAGCT      1       1586775,        1586865,        0.00    0.69
chr7    1587336 1587426 90      CTGTGCGCTCCGGGTGCCCCAGCTGGAAGGCAGGGTGTACCTGGTGAATTCTCCTTCTTCTCTGCGTACACCTGGCAGGGGAAGGCATAA      1       1587336,        1587426,        0.00    0.60
chr7    1587359 1587449 90      TGGAAGGCAGGGTGTACCTGGTGAATTCTCCTTCTTCTCTGCGTACACCTGGCAGGGGAAGGCATAACGCAGGGCCACGGAGGCGAACAG      1       1587359,        1587449,        0.00    0.59
chr7    1587404 1587494 90      CACCTGGCAGGGGAAGGCATAACGCAGGGCCACGGAGGCGAACAGCATCTCCACGCAGATGATGAAGTTCTGGTAGCCGGCGGCCAGCGT      1       1587404,        1587494,        0.00    0.63
chr7    1587449 1587539 90      CATCTCCACGCAGATGATGAAGTTCTGGTAGCCGGCGGCCAGCGTGCCAGCCCCCAGCTTGTTCCCGCCGCTGGTCTCCACCTCCGGGAT      1       1587449,        1587539,        0.00    0.66
chr7    1587494 1587584 90      GCCAGCCCCCAGCTTGTTCCCGCCGCTGGTCTCCACCTCCGGGATGACCCCGCACCGCTCCAGGATGGCCAGCAGCAGCCCTGCGGACGC      1       1587494,        1587584,        0.00    0.73
chr7    1587528 1587618 90      ACCTCCGGGATGACCCCGCACCGCTCCAGGATGGCCAGCAGCAGCCCTGCGGACGCCACGGCCGCTCAGCCCCAGCCCCAGACGGGGTCT      1       1587528,        1587618,        0.00    0.74
```

Here is the definition of each column of the *body* part.
* **chrom**, chromosome or super contig name;
* **start**, 0 based start coordinate of oligo covered region;
* **end**, 1 based end coordinate of oligo covered region;
* **seq_length**, the length of this oligo, please notice that there maybe some gaps in the oligo sequence compared with reference genome, so sequence length should be smaller or equal to the length of covered region. about how the gap comes in the oligo sequence, please refer to our algorithm part;
* **sequence**, sequence of this oligo, only A,C,G,T is allowed in the sequence
* **n_block**, block counts of this oligo, sometimes, oligo may cover several blocks in the reference genome;
* **starts**, 0 based start cooridinate of each blocks;
* **ends**, 1 based end cooridinate of each blocks;
* **repeat_ratio**, repeat ration of this oligo, the value is calculated by count the ratio of low case sequence in the reference sequence;
* **GC_content**, GC content ratio of this oligo;
* **rank**, rank of this oligo.


